fendermint/vm/interpreter/src/fvm/checkpoint.rs

```diff
@@ -322,23 +297,47 @@ fn should_create_checkpoint<DB>(
     gateway: &GatewayCaller<DB>,
     state: &mut FvmExecState<DB>,
     height: Height,
-) -> anyhow::Result<Option<checkpoint::SubnetID>>
+) -> anyhow::Result<Option<(Vec<checkpoint::IpcEnvelope>, checkpoint::SubnetID)>>
 where
     DB: Blockstore,
 {
-    if gateway.enabled(state)? {
-        let id = gateway.subnet_id(state)?;
-        let is_root = id.route.is_empty();
-
-        if !is_root && height.value() % gateway.bottom_up_check_period(state)? == 0 {
-            let id = checkpoint::SubnetID {
-                root: id.root,
-                route: id.route,
-            };
-            return Ok(Some(id));
-        }
+    if !gateway.enabled(state)? {
+        return Ok(None);
+    }
+
+    let id = gateway.subnet_id(state)?;
+    let is_root = id.route.is_empty();
+
+    if is_root {
+        return Ok(None);
+    }
+
+    let batch = gateway.bottom_up_msg_batch(state, height.into())?;
+
+    if batch.block_height.as_u64() != 0 {
```

Comment:
Have you considered some minimum batch size here?

Reply:
From the contract, there is no minimum batch size, it's either non-exists or some number of messages. If it's zero, that means the batch does not exists.

Reply:
I know, I meant: should it create a new checkpoint even if there is just 1 message in the batch, or should it wait until there is some, but not too many; e.g. make a checkpoint whenever there are at least 10 outgoing messages _or_ 1000 blocks have been created.